### DISCUSSION TITLE
Finer discretization than original topography scan

### DISCUSSION THREAD

**Discussion by sitangshugk95** [ContactEngineering/ContactMechanics#108] [Q&A] [answered]
In the current implementation, are we restricted to the discretization of the original topography scan? 
e.g., if the original scan had 256 x 256 pixes, is that the finest discretization at which we can solve the contact mechanics? Or could we solve it, for e.g., on 1024 x1024 pixels? 

In most FEM softwares for e.g., you can use the original scan to create the geometry and then discretize it however fine you want. Is there a way to do the same here? 

One way I can think of us is manually modifying the topography input file (by interpolation, albeit it will introduce some numerical artifacts) to have more data points - but I wanted to ask if such a functionality already exists.

**Comment by pastewka** [comment]
There is an `interpolate` and I think `fourier_interpolate` pipeline function that does just this. It may not be well documented though. Let me know if you have trouble finding documentation or using it.

So yes, it's possible to do this.

**Reply by sitangshugk95** [reply]
I was able to use it, but had a question about what it does to the power spectrum.

Here is the power spectrum, using `power_spectrum_from_profile()`, for the coarser sample (101 x 101 grid) : 
![image](https://github.com/ContactEngineering/ContactMechanics/assets/108830413/1a7ed9cb-22ee-474f-9946-a4783527bbf9)

Here is the power spectrum after it was discretized finer (501 x 501) using `interpolate_fourier()` : 
![image](https://github.com/ContactEngineering/ContactMechanics/assets/108830413/cbc304a0-9f36-4aee-abe4-5934107ebad4)

Why is there this huge change in the PSD profile, and is this expected to affect the contact problem? It also looks like periodicity is introduced into the topography after the operation (`is_periodic()`), but it doesn't it seem to affect the results significantly (fractional area/pressure distributions etc.)

P.S. When zoomed in to the same scale as the first image, the behavior is as expected (a shift up at the low wavelengths):
![image](https://github.com/ContactEngineering/ContactMechanics/assets/108830413/40fb6080-486c-4e3e-af96-7d06009cf78f)

**Reply by pastewka** [reply]
The drop in PSD is expected. You are interpolating the topography, which means it is essentially flat at scales below your original (non-interpolated) grid size. In other words, Fourier interpolation means filling in the low-wavelength components of the PSD with zeros.

If you use the normal `interpolate` function, the code will you bicubic spline interpolation (but that may not work for periodic topographies). For the bicubic spline I would expect some for of structure in the PSD at low wavelengths.

**Reply by sitangshugk95** [reply]
I could not find the `interpolate()` function - did you mean the [`interpolate_linear()` ](https://github.com/ContactEngineering/SurfaceTopography/blob/03338baa21ba6da6e6e313d52b247a0c397e6599/SurfaceTopography/Uniform/Interpolation.py#L34) function? If not, could you please direct me to the correct function?

If yes, what command should I use to return a topography instance, so that I can extract the new heights on the finer discretized surface? I am not sure what exactly the `linear_interpolator_topography()`  returns.

FYI here is what I have been using:
```
topo_coarse = read_topography('XYZ.txt',physical_sizes = (11e-6,11e-6))
t = topo_coarse.interpolate_linear()
t(10e-6,10e-6)
```

P.S. I cannot use the `interpolate_bicubic()` function because my topography is non-periodic - _do all the interpolation schemes implemented here make the topography periodic after interpolation_?

**Reply by pastewka** [reply]
The bicubic should work without periodicity. This is a bug and there is an issue open for it.

**Reply by pastewka** [reply, marked as answer]
PS: You are right that there may be no ˋinterpolatè function. The others are the ones I was referring to.
